src/Windows/dll/local_cache.cpp

```diff
@@ -1,33 +1,377 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
+#include <algorithm>
+#include <cstring>
+#include <mutex>
+#include <windows.h>
+#include <wincrypt.h>
+#include <bcrypt.h>
+#include <fstream>
+#include <stdio.h>
+#include <string>
+#include <sstream>
+#include <time.h>
+#include <iomanip> 
+#include <iostream>
+#include <cstdio>
+#include <filesystem>
 
-#include "local_cache.h"
+#define NT_SUCCESS(Status)          (((NTSTATUS)(Status)) >= 0)
 
-//
-// The Windows implementation of the local cache is not yet complete.
-// All methods noop, so we'll never cache any data locally.
-//
+#define STATUS_UNSUCCESSFUL         ((NTSTATUS)0xC0000001L)
+
+#define MAX_RETRY					10000
+#define SLEEP_RETRY_MS				15
+
+constexpr uint16_t CACHE_V1 = 1;
+
+static std::wstring g_cache_dirname;
+
+static void throw_if(bool should_throw, const std::string& error)
+{
+    if (should_throw)
+    {
+        throw std::runtime_error(error);
+    }
+}
+
+static void throw_if(bool should_throw, const std::string& error, HANDLE file)
+{
+	if (should_throw)
+	{
+		CloseHandle(file);
+		throw std::runtime_error(error);
+	}
+}
+
+static void fthrow_if(bool should_throw, const std::string& error, HANDLE file)
+{
+	if (should_throw)
+	{
+		FindClose(file);
+		throw std::runtime_error(error);
+	}
+}
+
+
+struct CacheEntryHeaderV1 {
+    uint16_t version;   // The version of the cache header
+    time_t expiry;      // expiration time of this cache item
+};
+
+static void make_dir(const std::wstring& dirname)
+{
+	CreateDirectory(dirname.c_str(), NULL);
+	throw_if(GetLastError() == ERROR_PATH_NOT_FOUND && GetLastError() != ERROR_ALREADY_EXISTS, "Path not found");
+}
+
+static void init_callback()
+{
+	const DWORD buffSize = 65535;
+	
+	LPTSTR env_home = new TCHAR[buffSize];
```

Comment:
This leaks the env_home memory.

Reply:
Another example where a std::unique_ptr would help.

Reply:
Done